runelite-client/src/main/java/net/runelite/client/plugins/demonicgorilla/DemonicGorillaPlugin.java

```diff
@@ -0,0 +1,706 @@
+			{
+				WorldPoint projectileSourcePosition = WorldPoint.fromLocal(
+					client, projectile.getX1(), projectile.getY1(), client.getPlane());
+				for (DemonicGorilla gorilla : gorillas.values())
+				{
+					if (gorilla.getNpc().getWorldLocation().distanceTo(projectileSourcePosition) == 0)
+					{
+						gorilla.setRecentProjectileId(projectile.getId());
+					}
+				}
+			}
+		}
+	}
+
+	private void checkPendingAttacks()
+	{
+		Iterator<PendingGorillaAttack> it = pendingAttacks.iterator();
+		while (it.hasNext())
+		{
+			PendingGorillaAttack attack = it.next();
+			if (tickCounter >= attack.getFinishesOnTick())
+			{
+				boolean shouldDecreaseCounter = false;
+				DemonicGorilla gorilla = attack.getAttacker();
+				MemorizedPlayer target = memorizedPlayers.get(attack.getTarget());
+				if (target == null)
+				{
+					// Player went out of memory, so assume the hit was a 0
+					shouldDecreaseCounter = true;
+				}
+				else if (target.getRecentHitsplats().size() == 0)
+				{
+					// No hitsplats was applied. This may happen in some cases
+					// where the player was out of memory while the
+					// projectile was travelling. So we assume the hit was a 0.
+					shouldDecreaseCounter = true;
+				}
+				else if (target.getRecentHitsplats().stream()
+					.anyMatch(x -> x.getHitsplatType() == Hitsplat.HitsplatType.BLOCK))
+				{
+					// A blue hitsplat appeared, so we assume the gorilla hit a 0
+					shouldDecreaseCounter = true;
+				}
+
+				if (shouldDecreaseCounter)
+				{
+					gorilla.setAttacksUntilSwitch(gorilla.getAttacksUntilSwitch() - 1);
+					checkGorillaAttackStyleSwitch(gorilla);
+				}
+
+				it.remove();
+			}
+		}
+	}
+
+	private void updatePlayers()
+	{
+		for (MemorizedPlayer mp : memorizedPlayers.values())
+		{
+			mp.setLastWorldArea(mp.getPlayer().getWorldArea());
+			mp.getRecentHitsplats().clear();
+		}
+	}
+
+	@Subscribe
+	public void onHitsplat(HitsplatApplied event)
+	{
+		if (event.getActor() instanceof Player)
+		{
+			Player player = (Player)event.getActor();
+			MemorizedPlayer mp = memorizedPlayers.get(player);
+			if (mp != null)
+			{
+				mp.getRecentHitsplats().add(event.getHitsplat());
+			}
+		}
+		else if (event.getActor() instanceof NPC)
+		{
+			DemonicGorilla gorilla = gorillas.get(event.getActor());
+			Hitsplat.HitsplatType hitsplatType = event.getHitsplat().getHitsplatType();
+			if (gorilla != null && (hitsplatType == Hitsplat.HitsplatType.BLOCK ||
+				hitsplatType == Hitsplat.HitsplatType.DAMAGE))
+			{
+				gorilla.setTakenDamageRecently(true);
+			}
+		}
+	}
+
+	@Subscribe
+	public void onGameState(GameStateChanged event)
+	{
+		GameState gs = event.getGameState();
+		if (gs == GameState.LOGGING_IN ||
+			gs == GameState.CONNECTION_LOST ||
+			gs == GameState.HOPPING)
+		{
+			reset();
+		}
+	}
+
+	@Subscribe
+	public void onPlayerSpawned(PlayerSpawned event)
+	{
+		Player player = event.getPlayer();
+		memorizedPlayers.put(player, new MemorizedPlayer(player));
```

Comment:
I do not like the overhead of wrapping + tracking hitsplats etc per player all the time. Can it be limited to only if there are gorillas around? Or maybe by region?